### ISSUE TITLE
Force a generic type to be a numeric one.

### ISSUE THREAD

**Issue by MaleDong** [dotnet/roslyn#3391]
Sometimes we have to write a generic Sort way or some special functions with a generic type that must be of a numeric one. However we cannot say something like this:

    public XXX<T> where T:ValueType

Though I know that int, long ,double……these basic elements come from ValueType. However I cannot do that because ValueType doesn't directly support operator comparation. 

And If I make "T" forced to IComparer<T>, IComparable……, here "T" doesn't ONLY a numeric one, any struct or class implemented this interface, it will have the "capability" to be pushed in.

And what's more——If my generic method/class have "+","-","*","/", only numeric type can do that.

So, it would be better if a generic type can be forced to a numeric one, and supports all kinds of numeric type, the numeric type should also include complicated struct such as "BigInteger".

It would be better if Microsoft can leave some special interfaces, virtual methods for us to create "customized data formation" and it will also be reguarded as "numeric type" (Maybe "Complex Number" is the future)

**Comment by whoisj** [comment]
Well there is where T : struct but it doesn't give you everything you asked for.

**Comment by MaleDong** [comment]
@whoisj ：Yes that's just my problem——this only limits "T" is type of struct, however we cannot control whether the struct is of a numeric type that supports Sort, +,-,*,/……operators.....

**Comment by MaleDong** [comment]
And this is very similar to Issue (https://github.com/dotnet/roslyn/issues/3255), however mine is the extension for him——NOT only for simple elements, but customized elements (such as Complex Number, BigInteger……)——These structures are also numeric type, as well.

**Comment by AdamSpeight2008** [comment]
I'd prefer to have the more generalised traits (#129). As it would allow to specify the operator the type must have implemented. Also @MaleDong be aware that the "comparison operator" are overloadable those don't have to the meaning of comparison. The contextual meaning of them is down to the type that is implementing them to decide.

**Comment by MaleDong** [comment]
In fact, I suddenly have a thought——To decide whether a number not, Maybe Microsoft can check:

1) Whether a customer has implemented "operator override" such as "+","-","*","/". —— If yes, this can be reguarded as a normal "Number" (fits the generic type's limitation). 

So Let's just say:

        public class GeneralOperation<T> where T: Some Microsoft's Number Type
       { ……}

And if the type to support Sort, just let us make it implement IComparable/ICompare<T> ;)

**Comment by AdamSpeight2008** [comment]
@MaleDong What is it about a type that makes that type a number?

Overriding `- + * / ` isn't enough. Or it being with being a `struct`

For example [Quaternion](Quaternion) which is non-communitive eg ` A + B != B + A` or `A * B  != B * A`

Also maintain a list of types that are considered "numeric" in the compiler, means that if say Microsoft implement new numeric type in a library, you have to also update the compiler to include it in the list.

Even adding an attribute or interface to the existing "numeric" types is a breaking change.

So what is it about a type that makes it numeric?

**Comment by whoisj** [comment]
I suppose we could look for an `intrinsic` keyword to signify `byte`, `sbyte`, `short`, `int`, `long`, `ushort`, `ulong`, `float`, `double`, and `decimal`. Could be a rather useful thing when developing math libraries. Instead of always assuming everything will need to be `double` we have fun things like `Vector<T> where T : intrinsic`, `Matrix<T> : intrinsic`, etc. I suppose `primitive`, the problem is that "intrinsic" already includes `char` and "primitive" includes `string` and `bool`.

**Comment by MaleDong** [comment]
@whoisj：Your way is quite nice for numeric type.
However, what about if you can create new key words (something like "numberic" that will includes all of the basic elements that belong to number type like "int","double","decimal").

And I'm not sure what you mean by:

Instead of always assuming everything will need to be double we have fun things like Vector<T> where T : intrinsic, Matrix<T> : intrinsic, etc. 

In fact, "T" is a generic type and if you use "int", the result and the intellisense is always "int".

**Comment by AdamSpeight2008** [comment]
> @MaleDong wrote:
> In fact, "T" is a generic type and if you use "int", the result and the intellisense is always "int
Not true. If the `T` is constrained `T : IComparable<T>` inside that method I'll only see the methods available on the interface + those from the base type `object`

**Comment by dsaf** [comment]
This might have been half-implemented:

http://stackoverflow.com/questions/10951392/implementing-arithmetic-in-generics
> Unfortunately, the type constraint does not allow you to require that the type supports arithmethic operators. What I find interesting is that in the BCL source code for e.g. Int32 you will find a IArithmetic<Int32> interface in the inheritance list that is commented out. This is pure speculation on my part but if Microsoft enabled that interface in the BCL then you could perhaps specify IArithmetic<T> as a constraint to allow you to write your own generic classes with arithmetic.

**Comment by HaloFour** [comment]
I think that the best way to handle this is to have generic type constraints expanded to support specific static (as well as instance?) methods/members of specified signatures.  That would cover overloaded operators, including those that affect heterogeneous types.  Unfortunately as of now the primitive types don't actually specify those operator methods.  They don't have to as the compilers know how to emit proper IL to perform direct addition of the primitive types.  But to be usable generically in this fashion they would need the overloaded operator methods otherwise they wouldn't satisfy the constraint.  The compiler would then be emitting IL to perform the static method call, which is far from ideal, but the JIT compiler is already smart enough to optimize those calls away.

I can imagine that the syntax for such constraints would be absolutely nasty, though.

```cs
public static T3 GenericAdd<T1, T2, T3>(T1 x, T2 y)
    where T1 : struct,
    where T2 : struct,
    where T3 : struct,
    where T1.op_Addition(T1, T2) -> T3
{
    return x + y;
}
```

Anything else just doesn't seem like it would fit in with the design of generics, specifically that a single member is emitted with one specific body that can be applied to any set of generic type arguments.  Compile-time templates would be more flexible and could solve this problem but would pose other problems.

**Comment by whoisj** [comment]
> In fact, "T" is a generic type and if you use "int", the result and the intellisense is always "int".

I was referring to how developers are forced, today, to develop math libraries against a know type (usually `double`) instead of being able to use a generic type due the lack of `where` clause keywords. I was providing a use-case for your request.

@HaloFour I really like that, though I'd prefer something more elegant. Maybe...
```
public static T3 GenericAdd<T1, T2, T3>(T1 x, T2 y)
    where T1 : struct,
    where T2 : struct,
    where T3 : struct,
    where operator +(T1, T2) => T3
    where operator +(T2, T1) => T3
{
    return x + y;
}
```

**Comment by AdamSpeight2008** [comment]
Can you show an example function from this maths library *(that some mention)*.

**Comment by MaleDong** [comment]
@dsaf @whoisj ：Your ideas has started me up to a general way, in fact:

1) Define a general interface that includes all kinds of operator methods such as "Add","Minus","MultipledBy","Divide","BitAnd","BitOr"……Consider we should be compatible with the previous basic types such as "int", "float", "double"……, Maybe this new interface should inherit from IArithmetic. Let's call it IOperator:

    public interface IOperator<T>
    {
             T Add(T other);
             T Minus(T other); 
             …………
            T BitAnd(T other);
            T BitOr(T other);
     } 

2) Define a general abstract class implement the interface with all of these methods "virtual" (To leave the space for customized Number type).

3) You can write something like:

    public class GeneralOperator<T> where T:IArithmetic<T>/IOperator<T> // (You make "T" either inherit from     //IArithmetic or IOperator , IOperator can have many more basic operator methods to be implemented)
    {
              public static T operator + (T t1, T t2)
             {
                       //VS should allow us to write something like this following:
                       return t1+t2;
                       //In fact, t1+t2 will be "reguarded as"—— t1.Add(t2)
             }

              
    }

**Comment by MaleDong** [comment]
Well……Thanks for your advices!

In order to simpify this problem……, We temporaily limit our thoughts to the same type's operation.

1) Define an interface or more interfaces to declare the numberic operation:

    public interface INumericOperator<T>
       {
        /// <summary>
        /// Add operator
        /// </summary>
        T Add(T t);
        /// <summary>
        /// Minus operator
        /// </summary>
        T Minus(T t);
        /// <summary>
        /// Multiplication operator
        /// </summary>
        T MultipledBy(T t);
        /// <summary>
        /// Division operator
        /// </summary>
        T DividedBy(T t);
      }

2) Write a generic type wrapper class that implement this interface, with all the methods left as "virtual" (because some methods may be not implemented by all of these methods):

   public class GenericOperationWrapper<T> : INumericOperator<T>
       {
        /// <summary>
        /// The actual value that is stored in the current wrapper class.
        /// </summary>
        protected T ActualValue { get; set; }
        /// <summary>
        /// The operation number
        /// </summary>
        public GenericOperationWrapper(T value)
        {
            ActualValue = value;
        }
        /// <summary>
        /// Do "Add" operation.
        /// Do the same by writing operation overrload like -,*,/……And call the same methods
        /// </summary>
        public static GenericOperationWrapper<T> operator +(GenericOperationWrapper<T> t1, GenericOperationWrapper<T> t2)
        {
            return new GenericOperationWrapper<T>(t1.Add(t2.ActualValue));
        }

        public override string ToString()
        {
            return ActualValue.ToString();
        }
        #region Numeric Virtual Methods
        public virtual T Add(T t)
        {
            throw new NotImplementedException();
        }

        public virtual T Minus(T t)
        {
            throw new NotImplementedException();
        }

        public virtual T MultipledBy(T t)
        {
            throw new NotImplementedException();
        }

        public virtual T DividedBy(T t)
        {
            throw new NotImplementedException();
        }
        #endregion
      }

3) Implement your own customized class like this:
    public class IntNumericOperator : GenericOperationWrapper<int>
       {
           public IntNumericOperator(int number)
            : base(number)
           {

           }
           public override int Add(int t)
           {
                return base.ActualValue + t;
           }
       }

4) And now you can use something like:
            IntNumericOperator op1 = new IntNumericOperator(1);
            IntNumericOperator op2 = new IntNumericOperator(2);
            Console.WriteLine(op1+op2);

5) Notice that my customized struct must implement the interface INumericOperator<T>, and it will be reguarded to support "Numeric Operations". So what I want Microsoft do is (Maybe):
5.1) Check if the generic struct type has implemented the interface INumericOperator<T>
5.2) If yes, when he/she is using +,-,*,/, The CLR or VS's intellisense should also translate these operators to the actual methods and call them.
5.3) For the very basic elements (such as int, float, double, decimal……), do as usual.
5.4) For other situations, if two structs of the same type don't implement the assigned interface, and he/she has used "+","-","*","/", the compiler should intellisense him/her that you should either implement the static operators or just implement the interface.

**Comment by whoisj** [comment]
@MaleDong I do not think you can expect interfaces to support operators. It provides horrible performance. Operators are static methods where as interface methods are instance methods, requiring `this` and all kinds of other setup to be done.

I think the original discussion about supporting improved `where` clause capabilities is the right direction to go if this is to be supported.

**Comment by MaleDong** [comment]
@whoisj：The reason why I use interface is:
1) We can customized our own "Numeric Type".
2) Make a general operator class with all kinds of operations (such as +,-,*,/) .
Of course, Here I mean "+" will automatically use "GenericOperationWrapper" to wrap the customized struct. "this" just means the "GenericOperationWrapper" itself.
I know the reason why don't like that because of the conversion of some very basic elements such as "int","double","decimal"……. I think this is the core thing that may be recoded to avoid the "horrible performace".

And as what you said, could you please offer us an example (in persudo code) to explain how to use that in details about your thoughts? Thanks anyway!

**Comment by whoisj** [comment]
See my proposition for how generic constraints could work. I believe they'd produce optimal code as well given the current compiler logic. Notice that I do not actually care that the operators are performing math of any kind. Instead, I only ask the compiler to constrain the type to one which has defined the `+` operator as necessary.

```
public static T3 GenericAdd<T1, T2, T3>(T1 x, T2 y)
    where operator +(T1, T2) => T3
    where operator +(T2, T1) => T3
{
    return x + y;
}
```

In this case, the `GenericAdd` method takes two types as parameters, requires that they implement the `+` operator for each other, and result in a known third type which is unrestricted.

I could image this for any of the existing operators:

```
public struct Vector3<T1>
    where operator +(T1, T1) => T1
    where operator -(T1, T1) => T1
    where operator *(T1, T1) => T1
    where operator /(T1, T1) => T1
    where operator +=(T1, T1) => T1
    where operator -=(T1, T1) => T1
    where operator ==(T1, Int32)
    where operator ==(T1, Double)
{
    // implementation ...
}
```

**Comment by MaleDong** [comment]
@whoisj：It seems Vs should add a new function——Supporting operator limitations For Generic Type. Am I right?

**Comment by MaleDong** [comment]
@whoisj：
It would be better if we can make "=="'s 2nd parameter type to "T1", of course I can take your meaning that you can define any kind of overload operators with one or two parameters of the different kinds of types.

**Comment by gafter** [comment]
Related to #2147 and #2204.

**Comment by msedi** [comment]
You are always talking about scalar values. But what happens when we have large arrays and we want to add them? Currently having math operation on two very long arrays works only well with unsafe code. Every other implementation I have tested failed completely.

Here I also need to have a pointer on a generic type. It would work well if the numeric type would also support pointers.

**Comment by whoisj** [comment]
> But what happens when we have large arrays and we want to add them?

You're assuming that the types can be addressed by unsafe pointers. If the types are references types or contain references, they cannot be. How would you address that?

I've proposed a keyword, which could be added to structs to insure that they are `unsafe` friendly (and the compiler errors if they contain illegal types).

**Comment by SirCmpwn** [comment]
Registering interest for this feature

**Comment by whoisj** [comment]
> Registering interest for this feature

@SirCmpwn there's a "Subscribe" button on the side, which should allow you avoid needlessly posted to get update notifications.

**Comment by SirCmpwn** [comment]
I am familiar with this button. It doesn't register interest, it subscribes to emails. Sending a notification to the participants revitalizes the discussion and draws renewed attention to the issue for planning.

**Comment by jnm2** [comment]
> Registering interest for this feature

This leaves me feeling revitalized for sure! :D
